src/drivers/amd/renoir/acp_bt_dma.c

```diff
@@ -0,0 +1,446 @@
+		return -EINVAL;
+	}
+	if (config->scatter) {
+		tr_err(&acp_bt_dma_tr, "BT_DMA: scatter enabled, that is not supported for now!");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int acp_dai_bt_dma_pm_context_restore(struct dma *dma)
+{
+	//TODO
+	return 0;
+}
+
+static int acp_dai_bt_dma_pm_context_store(struct dma *dma)
+{
+	//TODO
+	return 0;
+}
+
+static int acp_dai_bt_dma_copy(struct dma_chan_data *channel, int bytes,
+			  uint32_t flags)
+{
+	struct dma_cb_data next = {
+		.channel = channel,
+		.elem.size = bytes,
+	};
+	notifier_event(channel, NOTIFIER_ID_DMA_COPY,
+			NOTIFIER_TARGET_CORE_LOCAL, &next, sizeof(next));
+	return 0;
+}
+
+static int acp_dai_bt_dma_probe(struct dma *dma)
+{
+	int channel;
+
+	if (dma->chan) {
+		tr_err(&acp_bt_dma_tr, "ACP I2S DMA: Repeated probe");
+		return -EEXIST;
+	}
+	dma->chan = rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0,
+			    SOF_MEM_CAPS_RAM, dma->plat_data.channels *
+			    sizeof(struct dma_chan_data));
+	if (!dma->chan) {
+		tr_err(&acp_bt_dma_tr, "I2S DMA: Probe failure, unable to allocate channel descriptors");
+		return -ENOMEM;
+	}
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		dma->chan[channel].dma = dma;
+		dma->chan[channel].index = channel;
+		dma->chan[channel].status = COMP_STATE_INIT;
+	}
+	atomic_init(&dma->num_channels_busy, 0);
+	return 0;
+}
+
+static int acp_dai_bt_dma_remove(struct dma *dma)
+{
+	if (!dma->chan) {
+		tr_err(&acp_bt_dma_tr, "DMA: remove called without probe, it's a no-op");
+		return 0;
+	}
+	rfree(dma->chan);
+	dma->chan = NULL;
+	return 0;
+}
+
+static int acp_dai_bt_dma_get_data_size(struct dma_chan_data *channel,
+					uint32_t *avail, uint32_t *free)
+{
+	uint64_t tx_low, curr_tx_pos, tx_high;
+	uint64_t rx_low, curr_rx_pos, rx_high;
+
+	if (channel->direction == DMA_DIR_MEM_TO_DEV) {
+		tx_low = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_BT_TX_LINEARPOSITIONCNTR_LOW);
+		tx_high = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_BT_TX_LINEARPOSITIONCNTR_HIGH);
+		curr_tx_pos = (uint64_t)((tx_high<<32) | tx_low);
+		*free = (curr_tx_pos - prev_tx_pos) > ACP_BT_FIFO_BUFFER_SIZE ?
+			(curr_tx_pos - prev_tx_pos)%ACP_BT_FIFO_BUFFER_SIZE :
+			curr_tx_pos - prev_tx_pos;
+		*avail = ACP_BT_FIFO_BUFFER_SIZE - *free;
+		prev_tx_pos = curr_tx_pos;
+	} else if (channel->direction == DMA_DIR_DEV_TO_MEM) {
+		rx_low = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_BT_RX_LINEARPOSITIONCNTR_LOW);
+		rx_high = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_BT_RX_LINEARPOSITIONCNTR_HIGH);
+		curr_rx_pos = (uint64_t)((rx_high<<32) | rx_low);
+		*free = (curr_rx_pos - prev_rx_pos) > ACP_BT_FIFO_BUFFER_SIZE ?
+			(curr_rx_pos - prev_rx_pos)%ACP_BT_FIFO_BUFFER_SIZE :
+			(curr_rx_pos - prev_rx_pos);
+		*avail = ACP_BT_FIFO_BUFFER_SIZE - *free;
+		prev_rx_pos = curr_rx_pos;
+	} else {
+		tr_err(&acp_bt_dma_tr, "ERROR: Channel direction Not defined %d", channel->direction);
+		return -EINVAL;
+	}
+	return 0;
+}
```

Comment:
One thing to be wary of here is that I'm assuming your DSP is 32bit and you are reading a 64bit counter here which may overflow during the read (as the 2 * 32 bit reads are not atomic). The SOF Intel CAVS timer driver has the same 64bit readback issue, we need to read it at least twice to make sure the lower word has not overflowed during the read back.

Reply:
These are the Registers  read from the HW block .They represent the data consumed.